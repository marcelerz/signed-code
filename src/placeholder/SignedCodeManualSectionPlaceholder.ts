/**
 * Placeholders for signed code.
 *
 * @format
 */

import { getManualSection } from "../SignedCode";
import SignedCodePlaceholder from "./SignedCodePlaceholder";

/**
 * Class to handle the manual sections as placeholder.
 */
export default class SignedCodeManualSectionPlaceholder extends SignedCodePlaceholder {
	private name: string;
	private code: string | undefined;
	private previousCode: string | undefined;
	private defaultCode: string | undefined;

	constructor(name: string, code?: string, defaultCode?: string) {
		super();
		this.name = name;
		this.code = code;
		this.defaultCode = defaultCode;
	}

	/**
	 * Gets the unique name of the manual section.
	 */
	getName(): string {
		return this.name;
	}

	/**
	 * Gets the code of the manual section, if any.
	 */
	getCode(): string | undefined {
		return this.code;
	}

	/**
	 * Returns the previous code that was in a previous file and copied over
	 * to this placeholder through the `setPreviousCode` method.
	 */
	getPreviousCode(): string | undefined {
		return this.previousCode;
	}

	/**
	 * Adds previous code from a previous generated file to overwrite default
	 * values.
	 */
	setPreviousCode(code: string): void {
		this.previousCode = code;
	}

	/**
	 * Returns the default code defined by the template.
	 */
	getDefaultCode(): string | undefined {
		return this.defaultCode;
	}

	toString(): string {
		return getManualSection(
			this.name,
			this.code || this.previousCode || this.defaultCode || "",
		);
	}
}
