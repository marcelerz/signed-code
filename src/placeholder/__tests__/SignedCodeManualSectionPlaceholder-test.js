/**
 * Test for the SignedCodePlaceholder module.
 *
 * @format
 */

// SUT
import SignedCodeManualSectionPlaceholder from "../SignedCodeManualSectionPlaceholder";

describe("signedCodeManualSectionPlaceholder", () => {
	it("gets correctly initialized and returns name", () => {
		const name = "init-code";

		const instance = new SignedCodeManualSectionPlaceholder(name);

		expect(instance.getName()).toStrictEqual(name);
		expect(instance.getCode()).toBeUndefined();
		expect(instance.getDefaultCode()).toBeUndefined();
		expect(instance.getPreviousCode()).toBeUndefined();
	});

	it("gets the code", () => {
		const name = "init-code";
		const code = "let thereB = 'light';";

		const instance = new SignedCodeManualSectionPlaceholder(name, code);

		expect(instance.getName()).toStrictEqual(name);
		expect(instance.getCode()).toStrictEqual(code);
	});

	it("gets the default code", () => {
		const name = "init-code";
		const code = "let thereB = 'light';";

		const instance = new SignedCodeManualSectionPlaceholder(name, null, code);

		expect(instance.getName()).toStrictEqual(name);
		expect(instance.getDefaultCode()).toStrictEqual(code);
	});

	it("sets and gets the previous code", () => {
		const name = "init-code";
		const code = "let thereB = 'light';";

		const instance = new SignedCodeManualSectionPlaceholder(name);
		instance.setPreviousCode(code);

		expect(instance.getName()).toStrictEqual(name);
		expect(instance.getPreviousCode()).toStrictEqual(code);
	});

	describe("generates the code", () => {
		it("uses the code", () => {
			const name = "init-code";
			const code = "let thereA = 'light';";
			const previousCode = "let thereB = 'light';";
			const defaultCode = "let thereC = 'light';";

			const instance = new SignedCodeManualSectionPlaceholder(
				name,
				code,
				defaultCode,
			);
			instance.setPreviousCode(previousCode);

			expect(instance.toString()).toMatchSnapshot();
		});

		it("uses the previous code", () => {
			const name = "init-code";
			const previousCode = "let thereB = 'light';";
			const defaultCode = "let thereC = 'light';";

			const instance = new SignedCodeManualSectionPlaceholder(
				name,
				null,
				defaultCode,
			);
			instance.setPreviousCode(previousCode);

			expect(instance.toString()).toMatchSnapshot();
		});

		it("uses the default code", () => {
			const name = "init-code";
			const defaultCode = "let thereC = 'light';";

			const instance = new SignedCodeManualSectionPlaceholder(
				name,
				null,
				defaultCode,
			);

			expect(instance.toString()).toMatchSnapshot();
		});

		it("uses the plain code", () => {
			const name = "init-code";

			const instance = new SignedCodeManualSectionPlaceholder(name);

			expect(instance.toString()).toMatchSnapshot();
		});
	});
});
