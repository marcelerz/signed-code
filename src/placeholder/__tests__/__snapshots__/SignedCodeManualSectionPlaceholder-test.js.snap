// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signedCodeManualSectionPlaceholder generates the code uses the code 1`] = `"// <Manual-Section START init-code>let thereA = 'light';// <Manual-Section END>"`;

exports[`signedCodeManualSectionPlaceholder generates the code uses the default code 1`] = `"// <Manual-Section START init-code>let thereC = 'light';// <Manual-Section END>"`;

exports[`signedCodeManualSectionPlaceholder generates the code uses the plain code 1`] = `"// <Manual-Section START init-code>// <Manual-Section END>"`;

exports[`signedCodeManualSectionPlaceholder generates the code uses the previous code 1`] = `"// <Manual-Section START init-code>let thereB = 'light';// <Manual-Section END>"`;
